settings:Tab three organisation settings into single.

shashank-23002 opened the pull request:
<!-- Describe your pull request here.-->
This PR combines the "Users" , "Deactivate Users" and "Invitations" sections into a single tabbed panel (called "Users") .  This helps to make it possible to access all the functionality for managing users in one place.

Previously , the "Users" , "Deactivate Users" and "Invitations" sections were displayed into different components making it a little tedious to manage the functionalities altogether.

Now these sections are displayed into a single "Users" component making it easier to navigate through all the sections and access all the functionality for managing users in one place.

Fixes: #26949

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
<img width="920" alt="Screenshot 2023-12-27 at 2 46 36 AM" src="https://github.com/zulip/zulip/assets/97145463/bd555f8c-c650-4203-a2e5-510a8f985b38">

<img width="922" alt="Screenshot 2023-12-27 at 2 46 51 AM" src="https://github.com/zulip/zulip/assets/97145463/b82d70a2-670b-4f6d-80a1-bdfbc9308f97">

<img width="922" alt="Screenshot 2023-12-27 at 2 47 06 AM" src="https://github.com/zulip/zulip/assets/97145463/64f5353c-f145-4739-be93-c2e5e96c1f1b">

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks! Please update your PR to pass tests.

alya commented:
We should also update the PR so that the tabs expand to fit the strings (e.g., "Deactivated users"). Please post screenshots demonstrating that this works in both English and Russian.

shashank-23002 commented:
Thanks @alya , I have updated the above PR for tabs to expand according to string length. 
Also added a screenshot demonstrating the changes for Russian language.

However I feel that there is some issue with the puppeteer tests as I have tried many hits and trials but they fail in either case.
I would like to discuss about them in the [CZO thread](https://chat.zulip.org/#narrow/stream/138-user-questions/topic/failing.20puppeteer.20test)

shashank-23002 commented:
> Thanks! Those screenshots look fine to me; please ask @sahil839 for a first-round review once you've figured out how to fix the tests.

I am trying to figure out various ways to fix the tests but it seems to be an old problem without a final solution.
Would be great if it could be discussed on [CZO](https://chat.zulip.org/#narrow/stream/49-development-help/topic/puppeteer.20on.20M1.20Mac.20.3F).

sahil839 commented:
I don't think it is the old problem that is being discussed in CZO and it seems related to your changes. I just tried clicking on the deactivate button in dev server and the modal doesn't open, and the test is also failing for the same reason.

shashank-23002 commented:
Thanks @sahil839 for reviewing. 
I will try to fix the issue but as you can see in the above screenshots , the deactivated users modal opens up as expected.
 
Can you please specify the result after clicking on the deactivated users tab?

sahil839 commented:
I am talking about the "Deactivate" button showed using the icon in a user's row in the "Users" tab. Please read the tests carefully and see what it is trying to test.

shashank-23002 commented:
Thank you so much @sahil839 for pointing that out. 
I have fixed all the failing tests according to the new UI. PTAL

sahil839 commented on web/styles/app_components.css:
```diff
@@ -306,9 +302,8 @@ input::placeholder {
-                text-overflow: initial;
-                white-space: initial;
-                vertical-align: middle;
+                min-width: 0;
```
While I cannot find any bugs due to these CSS changes, it would be better if you could avoid these changes here and add the required changes for tabs used in user section only like we do for other cases so that we do not have any unnoticed bug. If we think that these CSS rules are unnecessary and can be updated, we can do so in a separate PR.

shashank-23002 replied:
I understand the concern. I have updated the required changes.
Though , it would be a great idea to update the ```tab-switcher``` component to increase it's width according to the text rather than defining the concrete-width for every class wherever required.

Probably there should be a separate issue/PR to update these changes.

timabbott replied:
Yeah that seems like a good potential follow-up PR.

shashank-23002 replied:
Sure Tim , I will immediately create a new follow-up PR once this is succesfully reviewed.

sahil839 commented on web/src/settings_users.js:
```diff
@@ -467,6 +468,49 @@ export function set_up_humans() {
+            activate_user_section(key);
+        },
+    });
+    $("#admin-user-list .tab-container").append(userToggler.get());
```
There is no code to load the invitations data, so no invitations are shown on toggling to the "invitations" tab.

shashank-23002 replied:
I have updated the changes to load the invitations data.
Hope that works.

sahil839 reviewed (commented):
Posted some comments. The "invitations" section does not work at all, so you would need to fix it.

sahil839 commented on web/src/settings_users.js:
```diff
@@ -467,6 +469,50 @@ export function set_up_humans() {
+            }
+
+            if (key === "deactivated-users-admin") {
+                allow_sorting_deactivated_users_list_by_email();
```
Why is this function called here? It just returns a value whether to allow sorting by email or not and I don't see any reason to call this function here.

shashank-23002 replied:
Ohh noted , cleaned it .

sahil839 reviewed (commented):
I think we can use a better approach -
- We can rename existing `user_list_admin.hbs` file to `active_user_list_admin.hbs`.
- And then we can have a new template file probably named as `user_list_admin.hbs` which will use the templates of three sections by using parameter like `data-user-settings-section` on the topmost element for each section.
- The `callback` function for toggler should be as simple as possible which would first hide all the sections and then show the appropriate one using `data-user-settings-section`.
 
You can see #28537 as a reference.

shashank-23002 commented:
> I think we can use a better approach -
> 
> * We can rename existing `user_list_admin.hbs` file to `active_user_list_admin.hbs`.
> * And then we can have a new template file probably named as `user_list_admin.hbs` which will use the templates of three sections by using parameter like `data-user-settings-section` on the topmost element for each section.
> * The `callback` function for toggler should be as simple as possible which would first hide all the sections and then show the appropriate one using `data-user-settings-section`.
> 
> You can see #28537 as a reference.

I tried to go for above approach as well , but looked at a feedback from previous PR #26966 in which @timabbott asked for the changes in ```settings_users.js``` file itself without involving any additional new files. 
Presently , I have tried to simplify the toggler and removed the extra functions.
Will try to look for above stated approach if this doesn't work.

sahil839 commented on web/templates/settings/admin_tab.hbs:
```diff
@@ -13,9 +13,11 @@
 {{> user_list_admin }}
-
-{{> deactivated_users_admin }}
+{{else}}
```
I don't understand these changes. I was thinking of only having {{user_list_admin}} here and then user_list_admin.hbs include the three templates for three tabs for all type of users.

shashank-23002 replied:
This was done to ensure that for non-admins we don't need to show all the three tabs , the Users section would just be sufficient as it was before.

sahil839 replied:
Was this discussed anywhere? We currently show all three sections as different panels in left side to non-admins.

shashank-23002 replied:
Ohh my bad , didn't look at the "Show more" section. 
I have updated those changes.

sahil839 commented on web/src/settings_sections.js:
```diff
@@ -34,8 +34,7 @@ export function get_group(section) {
 
-        case "user-list-admin":
-        case "deactivated-users-admin":
+        case "user-panel-admin":
```
I think we should avoid renaming `user-list-admin` for now. Since we would include active and deactivated terms to denote different tab sections.

sahil839 reviewed (commented):
Posted some comments.

shashank-23002 commented on web/templates/settings/invites_list_admin.hbs:
```diff
@@ -1,4 +1,4 @@
-<div id="admin-invites-list" class="settings-section" data-name="invites-list-admin">
+<div id="admin-invites-list" class="settings-section" data-user-settings-section="invites-list-admin">
```
If I remove that extra div i.e ```<div class="user-section-container" data-user-settings-section="user-list-admin">``` , then I need to have some class (```#users-admin-panel .settings-section``` in this case) which can be used to hide the user section on toggle.
Presently I use ```user-section-container``` to hide.

sahil839 replied:
Yeah, we need some common class but we can name it like `user-settings-section` instead of `settings-section` because that is used for each setting section opened from left panel.

shashank-23002 replied:
Sure , updated .

shashank-23002 commented on web/templates/settings/user_list_admin.hbs:
```diff
@@ -1,27 +1,17 @@
-            <tbody id="admin_users_table" class="admin_user_table"
-              data-empty="{{t 'No users match your current filter.' }}"></tbody>
-        </table>
+    <div class="user-section-container" data-user-settings-section="deactivated-users-admin">
```
I have removed the extra div. But the ```settings-section``` from each template file couldn't be removed because I need it to hide the users section in toggler callback.

sahil839 reviewed (commented):
Posted one comment. Also, there might be some unused code in `settings_sections.js` which should be removed now.

The commit message also needs some formatiing changes as per the [guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages). Also the commit message summary might be improved, "Tab" does not sound as a verb and make it somewhat confusing maybe using "Combine" would be better with some better phrasing to mention cleary what is being changed.

shashank-23002 commented:
Thanks @sahil839 , I have addressed all reviews. Hope that works.

sahil839 commented on web/templates/settings/active_user_list_admin.hbs:
```diff
@@ -0,0 +1,28 @@
+<div id="admin-active-users-list" class="user-settings-section" data-user-settings-section="user-list-admin">
```
I think we should also rename `data-user-settings-section` here to make it consistent with what we have for deactivated users template.

shashank-23002 replied:
Sure. 
Moreover I think we need to rename some endpoints as well.
Because for Deactivated Users , Invitations section initially were ```"/#organization/deactivated-users-admin"``` and ```"/#organization/invites-list-admin"```. But now we show all three sections using ```"/#organization/user-list-admin"```.(For reference - ```help_settings_links.py```).
Also we might need some documentation edits.

sahil839 replied:
Yes. I forgot about those. Maybe we can have different url for each tab like we do for right section in stream settings but am not sure. Would be good to discuss it on CZO.

shashank-23002 replied:
Currently, we show ```"/#organization/user-list-admin"``` for all the three tabs . But if we define the tab-url based on their keys then the active users section which opens by default would have ```\user-list-admin``` initially which will then be changed to ```\active-users-admin``` on tabbing. How do we need to sort this out ?
I have tried to changed the url on tabbing but the initial url still seems an issue.

sahil839 commented:
Posted one small comment. Also the "Fixes .." line in commit message does not end with a period.

Looks good otherwise but I am not sure about how to handle the banners. As per the current PR status, we show the banner for "Only organization administrators can edit these settings" to non-admins above the tab, but in main we do not show this banner in the invitations section. So, I guess would be better to show the banner below the tab and keep the behavior same as it was before. @alya thoughts?

GIF below for the current behavior implemented in this PR -
![Peek 2024-01-31 12-43](https://github.com/zulip/zulip/assets/35494118/aefb7947-9a6f-4864-901c-ef3c7396a49a)

alya commented:
Yeah, since the banners are tab-specific, having them just below the tabs makes sense to me.

alya commented:
I think the left-aligned tabs in the original screenshots are more consistent with other settings (Streams) than the center alignment in your latest screenshots. Is there a reason you decided to change it, or can we go back to the left alignment?

shashank-23002 commented:
@sahil839 fixed the url changes so you might want to have a look.

sahil839 commented:
The urls update correctly when I change the section by clicking on tabs but the correct tab is not opened when I change the url either manually or by using the browser back button.

Also, one browser back button bug that I can reproduce -
- Open the organization settings using gear menu.
- Click on "Users" in the left side. "Active users" tab is opened initially.
- Click on "Deactivated users" tab and then click on "Invitations" tab.
- Click on browser back button.

After clicking the browser back button, the expected url should be of the "Deactivated users" tab but instead the url changes to that of active users tab i.e. `#organization/user-list-admin/active-users-admin`.

To fix these you might probably need to have a look at code in `hashchange.js` and how we handle streams UI.

sahil839 commented:
Also, might be better to have the URL changes in a separate commit if sufficient code is added to fix the above mentioned bugs.

sahil839 commented:
We might also need to handle invalid URL cases. I have done some similar work in the "Handle invalid URLs" titled commits in #28764, but I am not sure about the approach used in that PR, so we can probably do this as follow-up.

sahil839 commented on web/src/settings_panel_menu.js:
```diff
@@ -151,7 +151,10 @@ export class SettingsPanelMenu {
-        const settings_section_hash = "#" + this.hash_prefix + section;
+        const settings_section_hash =
+            section === "user-list-admin"
+                ? `#${this.hash_prefix}${section}/active-users-admin`
```
The browser back button is not working correctly because the hash gets changed from `http://localhost:9991/#organization/user-list-admin/deactivated-users-admin` to `http://localhost:9991/#organization/user-list-admin/active-users-admin` here.

shashank-23002 replied:
Okay , so I have updated the code to update the hash for `user-list-admin` when the hash is set for each section based on data-section attribute ; i.e even before the activate_section_or_default function is called.
So this does not update the hash to `#organization/user-list-admin/active-users-admin` everytime the back button is clicked.

Though this works as we expect on pressing the back button , the section rendered doesn't change even though the url does change. Is this behaviour explainable? 
If so , should the probable solution be to always use the .goto() function for rendering a sub-section ?

sahil839 replied:
Can you update the PR with the changes you have done? I can take a look and we could debug what is preventing the UI to change.

shashank-23002 replied:
Sure,  I have updated the PR.

timabbott commented:
@shashank-23002 can you rebase this PR and post an update on what the status is of all outstanding feedback?

shashank-23002 commented:
Sure @timabbott ,  as per the current status - 

1) The Users section is currently combined of three sections (Users , Deactivated Users, Invitations) arranged in a tab format.
2) The banners have been properly aligned for each section as per [Alya's suggestions](https://github.com/zulip/zulip/pull/28359#issuecomment-1919691765).
3) The approach to render the section has been modified as per [Sahil's suggestions](https://github.com/zulip/zulip/pull/28359#pullrequestreview-1836203732) where we create a new template for Users section which contains the templates of "Active Users" , "Deactivated Users" and "Invitations". Then we render the tabbed section as per the `data-users-settings-section` attribute only after hiding all the sections.
4) Updated the hash for each section based on the key values (Default - `#organization/user-list-admin/active-users-admin`) so that clicking on each tab simultaneously changes the url for each section.
Also the browser back & forward buttons works as expected for each url to change as per the browser history.

**Remaining Concerns** - 
1) Although the url changes as expected , the section rendered still remains the same.
2) We cannot access a particular section from any part of the app using just the url. The section rendered after using any of three urls is the default ( Active Users ) section.

sahil839 commented:
> Remaining Concerns -
>
> 1. Although the url changes as expected , the section rendered still remains the same (Tried to figure out a solution using the popState event but did not completely solve the issue) .
> 2. We cannot access a particular section from any part of the app using just the url. The section rendered after using any of three urls is the default ( Active Users ) section.

You need to add code in `activate_section_or_default` to handle this. We already have code to show the panel based on what option in selected in the right side. Similarly, you need to add code to show the expected tab section based on the url.

shashank-23002 commented:
Thanks for the patience and advice @sahil839 . 
I think I have addressed the feedback as per your suggestions. 
It works well so you may want to have a look.

sahil839 commented:
One thing that is not working correctly is that if I type a url like `http://localhost:9991/#organization/user-list-admin/` then I expect the url to be changed to `http://localhost:9991/#organization/user-list-admin/active-users-admin`. Also if there is some invalid section after the `user-list-admin`, then as well the url should be changed to that of active users. We do this for the stream and group settings url.

Also, make sure that when updating the urls you use `history.replaceState` to make sure there is no entry for the invalid or incomplete url in browser history and the back button works correctly.

sahil839 commented on web/src/settings_panel_menu.js:
```diff
@@ -151,17 +159,34 @@ export class SettingsPanelMenu {
+                const newUrl = "#" + this.hash_prefix + section + "/active-users-admin";
+                history.replaceState({}, "", newUrl);
+            }
+            this.get_valid_users_section_panel(key);
```
I think this handling of hash can be done in `hashchange` because the invalid hash can only be used when manually typing urls or coming through a link with incorrect hash. And then this `activate_section_or_default` function can take a parameter like `right_section` which will only be passed if `section` is `user-list-admin`.

shashank-23002 replied:
@sahil839 I have tried to implement it taking examples from `streams` and `user groups` section. Hope this works.

sahil839 commented on web/src/settings_panel_menu.js:
```diff
@@ -53,6 +53,14 @@ export class SettingsPanelMenu {
+            */
+            if (section === "user-list-admin") {
+                browser_history.update_hash_internally_if_required(
+                    `#${this.hash_prefix}${section}/active-users-admin`,
```
Can this be removed by just passing `right_section = "active-users-admin"` to `activate_section_or_default` and then just update the hash in `activate_section_or_default` for all cases including the case when `section === "user-list-admin"`.

shashank-23002 replied:
Sure , updated. 
Though I am updating the hash in `get_valid_users_section_panel` rather than `activate_section_or_default` because it is possible to get the `right_section` as undefined in `activate_section_or_default` but it will never be the case for `get_valid_users_section_panel`.

sahil839 commented on web/src/settings_panel_menu.js:
```diff
@@ -151,17 +159,23 @@ export class SettingsPanelMenu {
 
+        if (section === "user-list-admin") {
+            right_section = right_section || window.location.hash.split("/")[2];
+            this.get_valid_users_section_panel(right_section);
```
Why is this needed? I guess we can assume that caller will make sure that the correct `right_section` variable is passed if `section == "user-list-admin"`.

shashank-23002 replied:
I have written this because initially when the hashchanged url is called , it triggers the `activate_section_or_default` function with the correct `right_section`. But then we call the `[data-user-settings-section].show()` function for rendering the correct section. Now this `show()` function again calls the `activate_section_or_default()` without a `right_section` property. Due to which `right_section` is again handled as `active-users-admin`.

sahil839 replied:
I am not sure whether you are talking about jquery `show` or `show()` function defined in `SettingsPanelMenu`. The `show()` function defined in `SettingsPanelMenu` is only called while toggling between personal and organization settings, so I guess we can pass the `right_section` there and it should be fine.

shashank-23002 replied:
Ohh yes that's right @sahil839 .
But then I could not find the reason why the `activate_section_or_default` function is called two times when I use hashchange to trigger it.
Due to this the first time , the call is made as expected but for the second time the right_section is found to be undefined which ultimately shows the `active-users-admin` section.

sahil839 commented on web/src/hashchange.js:
```diff
@@ -71,6 +71,20 @@ function is_somebody_else_profile_open() {
+        "invites-list-admin",
+    ]);
+    if (!valid_right_side_tab_values.has(right_section)) {
+        const newUrl = "#organization/user-list-admin/active-users-admin";
```
We prefer snake_case variable names.

sahil839 commented on web/src/hashchange.js:
```diff
@@ -323,7 +346,16 @@ function do_hashchange_overlay(old_hash) {
+                const right_section = handle_invalid_users_section_url(right_side_tab);
+                settings_panel_menu.org_settings.activate_section_or_default(
+                    section,
+                    right_section,
```
This can be better structured as -
```js
let right_section;
if (section === "user-list-admin") {
    const current_right_section = hash_parser.get_current_nth_hash_section(2);
    right_section = handle_invalid_users_section_url(right_side_tab);
}
settings_panel_menu.org_settings.activate_section_or_default(
    section,
    right_section,
);
```

sahil839 reviewed (commented):
Posted some comments.

sahil839 commented on web/src/settings_panel_menu.js:
```diff
@@ -123,7 +123,7 @@ export class SettingsPanelMenu {
     }
 
-    activate_section_or_default(section) {
+    activate_section_or_default(section, right_section = "active-users-admin") {
```
I don't think we should keep this default here since the right section value is only needed for one section. We can instead set it below only if `section === user-list-admin`.

shashank-23002 replied:
Ohh I guess we don't need to pass this , since we are passing it as a default for `get_valid_users_section_panel`.

shashank-23002 commented:
@sahil839 I have addressed the feedbacks. 

I am just not sure about how to address the ambiguity caused from `show()` function due to which I have to explicitly pass `right_section = window.location.hash.split("/")[2];`.

sahil839 commented:
Ohk, I looked at the code and I guess this is happening on calling `admin.launch` where we first call `activate_section_or_default` and then call `settings_toggle.highlight_toggle` which again calls `activate_section_or_default` in `show()`.

I guess we can refactor the code to not call `activate_section_or_default` in `admin.launch` and just set a field which will store the section name and then calling `current_tab` will return that section name.

shashank-23002 commented:
@sahil839 I have refactored the code and used the same approach to render the right panel.
Will fix the CI once the present changes of `curr_right_section` is finalized.

sahil839 commented on web/src/settings_panel_menu.js:
```diff
@@ -151,19 +151,25 @@ export class SettingsPanelMenu {
         this.$main_elem.children("li").removeClass("active");
         this.$curr_li.addClass("active");
-        this.curr_section = section
+        this.curr_section = section;
```
Make sure that these changes are done in the commit which added the code.

sahil839 commented on web/src/admin.js:
```diff
@@ -273,6 +273,6 @@ export function launch(section) {
 
     settings.open_settings_overlay();
-    settings_panel_menu.org_settings.activate_section_or_default(section);
+    settings_panel_menu.org_settings.curr_section = section || "organization-profile";
```
I think there is also a similar redundant call to `activate_section_or_default` in `hashchange.js` which we can be removed.

shashank-23002 replied:
> And the same changes should be done in settings.js as well.

You mean I need to do the refactoring for normal settings section as well? Should this be done in a separate commit?

sahil839 replied:
Yes. I think it can be done in either way - in the same commit or different commits. I guess same commit should work.

sahil839 commented on web/src/hashchange.js:
```diff
@@ -303,6 +317,12 @@ function do_hashchange_overlay(old_hash) {
+                right_section = handle_invalid_users_section_url(current_right_section);
+                settings_panel_menu.org_settings.curr_right_section = right_section;
+            }
             settings_panel_menu.org_settings.activate_section_or_default(section);
```
The `right_section` field can be passed here to `activate_section_or_default`. Basically we should just do what we do for `section`.

sahil839 commented on web/src/hashchange.js:
```diff
@@ -323,6 +343,12 @@ function do_hashchange_overlay(old_hash) {
+                right_section = handle_invalid_users_section_url(current_right_section);
+                settings_panel_menu.org_settings.curr_right_section = right_section;
+            }
             settings_panel_menu.org_settings.activate_section_or_default(section);
```
Here the `activate_section_or_default` call can be removed and the `curr_right_section` field should be set.

sahil839 commented on web/src/settings_panel_menu.js:
```diff
@@ -179,6 +185,16 @@ export class SettingsPanelMenu {
+
+        const $current_tab = $("#admin-user-list .tab-switcher .ind-tab");
+        $current_tab.removeClass("selected");
+        $(`#admin-user-list .tab-switcher .ind-tab[data-tab-key="${key}"]`).addClass("selected");
```
I think this can be removed once you do the changes I mentioned above in other comments.

shashank-23002 replied:
I have tried to do the changes in other comments but still this couldn't be removed.

sahil839 replied:
I guess we can directly call `toggler.goto` above by passing the `right_section` instead of having this function. `get_valid_users_section_panel` is mostly the code which is used in `goto`.

shashank-23002 replied:
Yeah definitely removing the `get_valid_users_section_panel` calls the toggler but displays all the three section. Even if i add a `toggler.goto` for first step , the main issue remains when we use `hashchange` to open the proper section. Because the toggler callback does not have any functionality to set the `selected` class for different tabs.

sahil839 replied:
You can just call `toggler.goto` and it will call `select_tab` and then `callback`. `select_tab` should handle `selected` class. I tried it and it works fine, but you should just test it once to make sure I did not miss any edge cases.

sahil839 reviewed (commented):
Posted few comments.

shashank-23002 commented:
@sahil839 I have addressed the comments and fixed the `hashchange.test.js` node test. The concerns remaining are - 

1) Dependency cycle within `settings_panel_menu.js` and `settings_users.js`.
2) Maybe setting `this.curr_right_section = "active-users-admin";` directly might not be a great idea but I can't find a very proper way for assigning this.

shubham-padia reviewed (commented):
@shashank-23002 You might need to rebase your PR to resolve conflicts. The conflicts arise from #29597, where we add roles filter to the activated users and deactivated users page. 

Also, in the first commit, I see `right_section` related code being added as part of the refactor, not sure if we should do that in a refactor when the right section functionality hasn't been reduced.

I've left some other suggestions you might wanna have a look at :) !

sahil839 commented on web/src/hashchange.js:
```diff
@@ -391,6 +416,14 @@ function do_hashchange_overlay(old_hash) {
+            if (settings_panel_menu.org_settings !== undefined) {
+                settings_panel_menu.org_settings.curr_right_section = right_section;
+            }
+        }
```
Yeah, I agree, defining function to set the sections would be nice.

sahil839 commented on web/src/hashchange.js:
```diff
@@ -321,9 +321,15 @@ function do_hashchange_overlay(old_hash) {
+            if (section === "user-list-admin") {
+                const current_right_section = hash_parser.get_current_nth_hash_section(2);
+                right_section = handle_invalid_users_section_url(current_right_section);
+                settings_panel_menu.org_settings.curr_right_section = right_section;
```
This right section code should be in the next commit.

sahil839 reviewed (commented):
Posted a couple of comments.
